File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.cc
#include <iostream>

using namespace std;

int tri[501][501], dp[501][501], n;

void input() {
	cin >> n;
	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= i; j++)cin >> tri[i][j];
	}
}

void solution() {
	int result = -1;
	dp[1][1] = tri[1][1];
	for (int i = 2; i <= n; i++) {
		for (int j = 1; j <= i; j++)dp[i][j] = max(dp[i - 1][j - 1], dp[i - 1][j]) + tri[i][j];
	}
	for (int i = 1; i <= n; i++) result = dp[n][i] > result ? dp[n][i] : result;
	cout << result;
}

int main() {
	ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
	input();
	solution();
}

File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.py
n = int(input())

tri = [list(map(int, input().split(' '))) for _ in range(n)]
dp = [[0]*len(tri[i]) for i in range(n)]

dp[-1] = tri[-1]

for i in range(n - 2, -1, -1):
    for j in range(len(tri[i])):
        dp[i][j] = max(dp[i+1][j], dp[i+1][j+1]) + tri[i][j]
print(*dp[0])
